### PR TITLE
[SPARK-8884][MLlib] 1-sample Anderson-Darling Goodness-of-Fit test

### PR DESCRIPTION
**Pull request by hhbyyh**
## What changes were proposed in this pull request?

jira: https://issues.apache.org/jira/browse/SPARK-8884
This is a continued work of @josepablocam in https://github.com/apache/spark/pull/7278. 

As sync in jira, https://issues.apache.org/jira/browse/SPARK-8884, I'll help with the rest work. I've resolved the conflict, updated links and rewrote the implementation following the original formula. https://en.wikipedia.org/wiki/Anderson%E2%80%93Darling_test. A new ut is also added for weibull and gumbel distribution.

We have implemented one-sample Anderson-Darling goodness-of-fit test to add to the current hypothesis testing functionality. The current implementation supports various distributions (normal, exponential, gumbel, logistic, and weibull). However, users must provide distribution parameters. In contrast to other tests, such as the Kolmogorov Smirnov test, we only support specific distributions as the critical values depend on the distribution being tested.
## How was this patch tested?

unit tests and some manual comparison with R and Python.

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/mllib/stat/test/AndersonDarlingTest.scala [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+    : Iterator[(Double, Double, Double)] = {
+    val initAcc = (0.0, 0.0, 0.0)
+    val pResult = part.zipWithIndex.foldLeft(initAcc) { case ((prevS, prevC, prevCt), (v, i)) =>
+      val y = dist.cdf(v)
+      val a = math.log(y)
+      val b = math.log(1 - y)
+      val unAdjusted = a * (2 * i + 1) + b * (2 * n - 2 * i - 1)
+      val adjConstant = a - b
+      (prevS + unAdjusted, prevC + adjConstant, prevCt + 1)
+    }
```
Comment (hhbyyh) [inline mllib/src/main/scala/org/apache/spark/mllib/stat/test/AndersonDarlingTest.scala]: I tried to add an ut for gumbel distribution. yet the number I get from scipy is different from the value calculated here. I'm not sure if the expression here is effective for the asymmetrical distribution. I will investigate more and any comment will be welcome.

**Comment by hhbyyh** [comment]
building error:
/home/jenkins/workspace/SparkPullRequestBuilder/mllib/src/main/scala/org/apache/spark/mllib/stat/test/AndersonDarlingTest.scala:26: object Logging is not a member of package org.apache.spark

need a retest

**Comment by mengxr** [comment]
@hhbyyh We moved Logging to `org.apache.spark.internal`. Could you update this PR?

**Comment by hhbyyh** [comment]
@mengxr thanks. This is ready for review.

**Comment by hhbyyh** [comment]
@mengxr ping since this is targeting 2.0

**Comment by hhbyyh** [comment]
Close this.

**Pull request closed without merging**